### PR TITLE
feat(developer): kmc-copy

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes: #2264.

Basic functionality for cloning and renaming a keyboard project with the `kmc copy` command. Many TODO-COPY items are resolved in #12586. Others to come.

**Note:** TODO list moved to #12641.

# User Testing

* Install Keyman Developer
* Open a command prompt

* **TEST_COPY_KEYBOARDS:** Try to copy an existing keyboard project with the `kmc copy` command: `kmc copy <source-project.kpj> -o <destination-folder>`, e.g. `kmc copy keyboards/release/k/khmer_angkor -o c:\temp\my_first_cambodian_keyboard`. Do this with a few different keyboards. Verify that the copy and rename worked by building the new keyboards and verify that they build. Verify that the filenames are as expected in the new keyboards.
* **TEST_COPY_MODELS:** Try to copy an existing lexical model project with the `kmc copy` command: `kmc copy <source-project.kpj> -o <destination-folder>`, e.g. `kmc copy lexical-models/release/sil/sil.km.cnd -o c:\temp\mydict.km.cambodian`. Do this with a few different lexical models. Verify that the copy and rename worked by building the new models and verify that they build. Verify that the filenames are as expected in the new models.

### CODE REVIEW

**Comment by dinakaranr** [comment]
# Test Results
I tested this issue with the attached "18.0.127-alpha-test-12555" build on the Windows 10 and 11 OS. Here is my observation.

* **TEST_COPY_KEYBOARDS (Passed):** 
1. I installed the keymandeveloper-18.0.127.exe build in the Windows 10.
2. Navigate to the folder path that has GitHub's keyboard.
3. Open the command prompt.
4. Enter the below command in the window's prompt
5. >kmc copy C:Users\Bharani\Documents\GitHub\keyboards\release\gautami\gautami_thamizh_tamil\gautami_thamizh_tamil.kpj -o C:\Users\Bharani\AppData\Local\Temp\KMCTesting1
6. >kmc copy C:\Users\Bharani\Documents\GitHub\keyboards\release\gautami\gautami_thamizh_tamil -o C:\Users\Bharani\Downloads\KMCTesting3
7. Verified that the keyboard folder or *.kpj file was copied into a new folder. 
8. Verified that the keyboard file was renamed in the new folder.

* **TEST_COPY_MODELS (Passed):** 
1. I installed the keymandeveloper-18.0.127.exe build in the Windows 10.
2. Navigate to the folder path that has GitHub's keyboard(lexical-models).
3. Open the command prompt.
4. Enter the below command in the window's prompt
5. >kmc copy C:\Users\Bharani\Documents\GitHub\lexical-models\release\sil\sil.cmo.bw -o D:\Dinakaran_Files\TestingKeyboard\LexicalModelKMCTesting1
6. lexical-models
7. >kmc copy C:\Users\Bharani\Documents\GitHub\lexical-models\release\sil\sil.cmo.bw\sil.cmo.bw.kpj -o D:\Dinakaran_Files\TestingKeyboard\LexicalModelKMCTesting
8. Verified that the keyboard folder or *.kpj file was copied into a new folder. 
9. Verified that the keyboard file was renamed in the new folder.
It works well. Thanks.

- I observed two things for copying files.
- I created a folder in the Temp folder and then tried to copy a keyboard file.
- Here, the folder "already exists, not overwriting" message appears on the command prompt.
- I created a folder in the "Download" folder and then tried to copy a keyboard *.kmp file.
- Here, folder "Could not find project file" message appears on the command prompt.
Please refer to the attached screenshot

File: developer/src/kmc-copy/src/KeymanProjectCopier.ts [hunk 1]
```diff
@@ -0,0 +1,561 @@
...
+    };
+
+    return true;
+  }
+
+  private copySourceFile(project: KeymanDeveloperProject, filename: string, outputPath: string, source: string, result: CopierResult): boolean {
+    return this.copyGenericFile(project, filename, this.callbacks.path.join(this.outPath, 'source'), source, result);
+  }
+
+  private copyModelTsSourceFile(project: KeymanDeveloperProject, filename: string, outputPath: string, source: string, result: CopierResult): boolean {
```
Comment (darcywong00) [inline developer/src/kmc-copy/src/KeymanProjectCopier.ts:277]: So this also copies tsv wordlists?

Comment (mcdurdin) [reply developer/src/kmc-copy/src/KeymanProjectCopier.ts:277]: Yes, just like `copyKmnSourceFile` copies icons, etc. Each function is responsible for copying the files that are referenced by its input file.

File: developer/src/kmc-copy/src/KeymanProjectCopier.ts [hunk 2]
```diff
@@ -0,0 +1,561 @@
...
+            }
+
+            // Rewrite keyboard id
+            if(keyboard.ID == this.sourceId) {
+              keyboard.ID = this.outputId;
+            }
+          }
+        }
+
+        if(kps.Package?.LexicalModels?.LexicalModel?.length) {
```
Comment (darcywong00) [inline developer/src/kmc-copy/src/KeymanProjectCopier.ts:387]: Are we still flagging an error if a kps includes both keyboard and lexical model?

File: developer/src/kmc-copy/test/fixtures/projects/alephwithbeth/build/alephwithbeth.keyboard_info [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    "linux": "full",
+    "desktopWeb": "full",
+    "ios": "full",
+    "android": "full",
+    "mobileWeb": "full"
+  },
+  "minKeymanVersion": "9.0",
+  "sourcePath": "experimental/a/alephwithbeth",
+  "helpLink": "https://help.keyman.com/keyboard/alephwithbeth"
+}
```
Comment (darcywong00) [inline developer/src/kmc-copy/test/fixtures/projects/alephwithbeth/build/alephwithbeth.keyboard_info:41]: Confirming you wanted to check in all the test/fixtures/ build/ artifacts?

File: developer/src/kmc-copy/src/KeymanProjectCopier.ts (hunk 2 above)
Comment (mcdurdin) [reply developer/src/kmc-copy/src/KeymanProjectCopier.ts:387]: Only in the compiler, not in the copying tool. If there are problems in the source project, they'll be replicated in the target project.

File: developer/src/kmc-copy/test/fixtures/projects/alephwithbeth/build/alephwithbeth.keyboard_info (hunk 3 above)
Comment (mcdurdin) [reply developer/src/kmc-copy/test/fixtures/projects/alephwithbeth/build/alephwithbeth.keyboard_info:41]: Yes, these are needed for these tests.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.138-alpha](https://keyman.com/downloads/releases/alpha/18.0.138)

**Pull request merged**